Add metrics to identify audit logging failures

It would be very helpful to be able to alert an operator when there are audit logging failures. Because this information isn't available from the /sys/health endpoint, we need some other means.

I would suggest the addition of some appropriate metrics in the telemetry, so alerting can be done from statsd/statsite.

I don't have strong feelings about exactly what the metrics should be. Being able to monitor 500 response codes would certainly help, or maybe the audit logging backends should provide more specific error metrics.

See also [this conversation](https://groups.google.com/forum/#!topic/vault-tool/tnLNRNERdvs).

I just had this issue in production this morning.

This morning we had a failure in our audit backend(the file was rotated, but the -HUP signal did not happen for some reason).

vault log:

2017/06/19 05:41:59.431861 [ERROR] audit: backend failed to log response: backend=file/ error=write vault_audit.log: bad file descriptor
2017/06/19 05:41:59.431874 [ERROR] core: failed to audit response: request_path=auth/token/renew-self error=no audit backend succeeded in logging the response2017/06/19 05:42:26.046577 [ERROR] audit: backend failed to log request: backend=file/ error=write vault_audit.log: bad file descriptor
2017/06/19 05:42:26.046605 [ERROR] core: failed to audit request: path=auth/token/renew-self error=1 error occurred:
* no audit backend succeeded in logging the request

yet, the health checks were continuing to pass, despite error 500's on any read/write call to vault.  This seems.. wrong somehow :)

Ideally, I'd love if vault on a bad FD would just try a re-open like if a HUP signal happened. Especially since audit is crucial to a valid operating vault system.

also, the /health check should probably at least WARN, if not outright FAIL, if it can't write the audit log for whatever reason.

--- a/vault/audit.go
+++ b/vault/audit.go
@@ -507,6 +507,10 @@ func (a *AuditBroker) LogRequest(auth *logical.Auth, req *logical.Request, heade
 		}
 
 		ret = retErr.ErrorOrNil()
+
+		if ret != nil {
+			metrics.IncrCounter([]string{"audit", "log_request_failure"}, 1.0)
+		}
 	}()
 
 	// All logged requests must have an identifier
@@ -565,6 +569,10 @@ func (a *AuditBroker) LogResponse(auth *logical.Auth, req *logical.Request,
 		}
 
 		ret = retErr.ErrorOrNil()
+
+		if ret != nil {
+			metrics.IncrCounter([]string{"audit", "log_response_failure"}, 1.0)
+		}
 	}()
 
 	headers := req.Headers

